### ISSUE TITLE
WC_Order::$status_transition must be emptied if something prevent save() from happening

### ISSUE THREAD

**Issue by axi** [woothemes/woocommerce#46886]
### Prerequisites

- [x] I have carried out troubleshooting steps and I believe I have found a bug.
- [x] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

Using `'woocommerce_before_' . $this->object_type . '_object_save'` to prevent saving of an Order does not prevent the  `woocommerce_order_status_X` hooks to be called

### Expected behavior

If we throw an Error that prevents changes, no further status change hook must be called.

### Actual behavior

In one of my devs, I need to prevent some order change status from happening (even the admin is not authorised to do so manualy from the backend) but the existing order admin page still give this availability in several places:
- the status selector 
![image](https://github.com/woocommerce/woocommerce/assets/389340/ed359640-0d4f-4d1d-a5c8-d29b3b94a791)
- the buttons in preview view
![image](https://github.com/woocommerce/woocommerce/assets/389340/bee35349-eecc-4e3e-8a4b-dee7cb20c98a)

Even if I could remove those button, I still need to prevent those changes on a low level.

Currently, hitting one of those buttons trigger a `WC_Order::set_status()`, filling the `WC_Order::$status_transition` values  and that's fine.

I use the `WC_Abstract_Order::save()` `'woocommerce_before_' . $this->object_type . '_object_save'` action hook to check some conditions and  I throw an Exception that is catched in this same function.

At this point, it cancel the changes (no save) but  `WC_Order::$status_transition` still holds the changes and because `WC_Order::status_transition()` is still called (in `WC_Order::save()`), lot of unexpected hooks get called.

To resume, if I prevent saving a status change by using `'woocommerce_before_' . $this->object_type . '_object_save'`  hook, none of the `woocommerce_order_status_X`hooks should be called.

### Steps to reproduce

```php
add_action(
    hook_name: 'woocommerce_before_order_object_save',
    callback: static function (WC_Order $order) {
        $changes = $order->get_changes();
        if (!empty($changes) && isset($changes['status']) && $changes['status'] === 'refunded') {
            throw new \RuntimeException('to prove my point');
        }
    }
);

add_action(
    hook_name: 'woocommerce_order_status_refunded',
    callback: static function ($id) {
        die('I should have never been called');
    }
);

$order = new Wc_Order();
$order->set_status('draft');
$order->save();
$order->set_status('refunded');
$order->save();
```

### WordPress Environment

WooCommerce version: 		8.8.2
WordPress version: 		6.5.2

### Isolating the problem

- [ ] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/products/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

**Comment by vedanshujain** [comment]
@axi going through the PR and issue, it seems like this might be a bit invasive, at least for the use case you mention and will have unintended side effects. For example, what if an exception is coming from the `after` hook instead of before, and the order is indeed saved and the status should be transitioned? 

For your use case, we check whether the user has permission to edit, for example, by checking `current_user_can( 'edit_post', $post_id )` so you can probably use the `user_has_cap` to remove the edit_shop_order capability when order id is passed in one of the args?

cc @naman03malhotra

**Comment by axi** [comment]
@vedanshujain you're right, I didn't think about this but I provided an update to the PR.
I'll check you proposal for an alternate way to fix my usecase but this still is a bug that should be fixed, isn't it ?

**Comment by axi** [comment]
PS: I checked my usecase and and found that preventing a save action in this central location was probably the best thing to do.
WC_Abstrac_Order::save() method is called in tens of location and I would have to check my conditions in many places to have the same effect.

![image](https://github.com/woocommerce/woocommerce/assets/389340/3670e72a-5565-43f0-9c1b-453e97a4e434)
